Feat/alerts scoped to env

marcambier opened the pull request:
**Issue**

https://github.com/gravitee-io/issues/issues/6079

**Description**

feat: scope platform alerts to environment

- Add a environment_id on each alert_trigger
- Platform alerts are scoped to environment, with referenceType=ENVIRONMENT & referenceId=environment.id

<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-kepzbijuwt.chromatic.com)
<!-- Storybook placeholder end -->
<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/feat-alertsscopedtoenv-master/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
<!-- UI placeholder end -->

Commit:
feat: scope platform alerts to environment

- Add a environment_id on each alert_trigger
- Platform alerts are scoped to environment, with referenceType=ENVIRONMENT & referenceId=environment.id

a-cordier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java:
```diff
@@ -357,10 +353,10 @@ public AlertTriggerEntity apply(AlertTrigger alertTrigger) {
                 throw new AlertNotFoundException(alertId);
             }
-            final AlertTriggerEntity alert = convert(optionalAlert.get());
+            final AlertTriggerEntity alert = alertTriggerConverter.toAlertTriggerEntity(optionalAlert.get());
```
maybe even

```java
final AlertTriggerEntity alert = alertTriggerRepository.findById(alertId)
    .filter(a -> a.getReferenceId().equals(referenceId))
    .map(alertTriggerConverter::toAlertTriggerEntity)
    .orElseThrow(() -> new AlertNotFoundException(alertId));
```

marcambier replied:
nice, thanks :+1:

gcusnieux commented on .circleci/config.yml:
```diff
@@ -290,6 +290,9 @@ jobs:
                   command: |
+                      cd gravitee-apim-repository/gravitee-apim-repository-api
+                      mvn -s ../../.gravitee.settings.xml clean install -DskipTests
+                      cd ../..
```
```
- run:
      name: Build repository models
      command: |
            mvn clean install -s .gravitee.settings.xml 
            -f gravitee-apim-repository/gravitee-apim-repository-api  
            -Dskip.validation=true -DskipTests
- run:
      name: Run tests
      command: |
            mvn -pl 'gravitee-apim-definition, gravitee-apim-rest-api, gravitee-apim-gateway' -amd 
            --fail-fast -s .gravitee.settings.xml test --no-transfer-progress 
            -Dskip.validation=true -T 2C
```

marcambier replied:
This is not related to this PR.
Will apply your remark on a dedicated PR.

gcusnieux commented on gravitee-apim-console-webui/src/management/management.route.ts:
```diff
@@ -127,8 +127,9 @@ function managementRouterConfig($stateProvider) {
+          AlertService.listAlerts(undefined, Scope.ENVIRONMENT, false).then((response) => response.data),
+        alertingStatus: (AlertService: AlertService) =>
+          AlertService.getStatus(undefined, Scope.ENVIRONMENT).then((response) => response.data),
       },
```
Not related to your changes, but AlertService signatures could be refactored to pass the undefined as an optional parameter

marcambier replied:
Yes, I refactored that part in a dedicated commit :+1:

gcusnieux commented on gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertResource.java:
```diff
@@ -83,14 +83,18 @@ public Response updateAlert(
 
     private void checkPlugins() {
-        AlertStatusEntity alertStatus = applicationAlertService.getStatus();
+        AlertStatusEntity alertStatus = applicationAlertService.getStatus(GraviteeContext.getExecutionContext());
```
Could you pass the execution context as a parameter to instantiate it only once ?

marcambier replied:
:+1:

gcusnieux commented on gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/ApplicationAlertsResource.java:
```diff
@@ -119,7 +119,7 @@ public ApplicationAlertResource getApplicationAlertResource() {
 
     private void checkPlugins() {
-        AlertStatusEntity alertStatus = applicationAlertService.getStatus();
+        AlertStatusEntity alertStatus = applicationAlertService.getStatus(GraviteeContext.getExecutionContext());
```
here also

marcambier replied:
:+1:

gcusnieux commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/AlertTriggerConverter.java:
```diff
@@ -0,0 +1,180 @@
+
+    public AlertTrigger toAlertTrigger(ExecutionContext executionContext, final UpdateAlertTriggerEntity alertEntity) {
+        final AlertTrigger alert = new AlertTrigger();
+        alert.setId(UUID.toString(UUID.random()));
```
This should not be `alert.setId(alertEntity.getId())` ?

marcambier replied:
Yes.
It was already like that in the mapping before refactoring.
But it's nonsense, the caller was setting back the id using the one from UpdateAlertTriggerEntity.
I changed it. Nice catch.

gcusnieux commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java:
```diff
@@ -152,45 +152,48 @@ public class AlertServiceImpl extends TransactionalService implements AlertServi
 
     @Override
-    public AlertStatusEntity getStatus() {
+    public AlertStatusEntity getStatus(ExecutionContext executionContext) {
```
We could use final ExecutionContent on each signatures

gcusnieux commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AlertServiceImpl.java:
```diff
@@ -410,102 +403,48 @@ public AlertEventEntity apply(AlertEvent alertEventRepo) {
+                                    triggerEntity.setEventRules(null);
+                                    triggerEntity.setParentId(alertId);
+                                    triggerEntity.setCreatedAt(new Date());
+                                    triggerEntity.setUpdatedAt(trigger.getCreatedAt());
```
Not update by you, but it's strange no ?

marcambier replied:
Yes.
The updatedAt takes the creation date of the template (so, update date is before creation date).
I Can't explain why.

Commit:
refactor(console): improve readability of alertService listAlerts and getStatus methods

- uses optional parameters instead of calls with undefined
- callers uses enums instead of integers

ThibaudAV reviewed (approved):
LGTM 👏 👏 

Test locally and sounds good :)
